File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int>mp;
        int i=0,j=0,n=s.length();
        int ans=INT_MIN;
        while(j<n){
            if(mp[s[j]]>0){
                mp[s[i]]--;
                i+=1;
            }
            else{
                mp[s[j]]++;
                j+=1;
                ans=max(ans,j-i);
            }
        }
        return ans==INT_MIN?0:ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character,Integer>mpp=new HashMap<>();
        int n=s.length();
        int l=0,r=0;
        int ans=0;
        while(r<n){
            if(mpp.containsKey(s.charAt(r))) l=Math.max(mpp.get(s.charAt(r))+1,l);
            mpp.put(s.charAt(r),r);
            ans=Math.max(ans,r-l+1);
            r+=1;
        }
        return ans;
    }
}